Add --check-untyped-defs to mypy.ini and test suite

rgommers opened the pull request:
Also move the ignores for imports to the bottom in `mypy.ini`, those are much less interesting - start with the stuff people want to work on.

Second commit tests the instructions: remove an ignore, fix the issue.

rgommers committed:
Document how to run and improve type annotations better.

Also move the ignores for imports to the bottom in mypy.ini,
those are much less interesting - start with the stuff people
want to work on.

rgommers committed:
Remove `ignore_errors` from mypy.ini for torch.serialization

ntenenz commented on mypy-README.md:
```diff
@@ -11,3 +12,53 @@ mypy
+2. Inline type annotations for all Python code, _except_ if there are too many
+   overloads for functions/methods - in that case a stub file should be
+   preferred (what's too many is a bit of a judgement call, I'd suggest two per
+   function is a reasonable threshold).
```
One of the benefits of type hints is that they are self-documenting. For this reason, I'd personally be a bit more aggressive in the number of permissible `@overload`s permitted before going to stubs. Unfortunately, I don't have a firm number in mind.

If feasible, what might be beneficial is an analysis of the distribution of `@overload`s per function/method to inform the decision.

rgommers replied:
Right now I don't see any that exceed the threshold. `torch/nn/modules/container.pyi` has some double overloads (`__init__`, `__getitem__`). I expect the number of overloads to grow significantly though when people want to make signatures more specific - things like `Union[Iterable[Tensor], Iterable[dict]]` and `Union[int, List[int], Size]` may turn into overloads.

That said, point taken, I'll bump it to 3.

ezyang replied:
#35566 is another case where we may not be able to put things in code, and have to move it to stubs.

ntenenz commented on mypy-README.md:
```diff
@@ -11,3 +12,53 @@ mypy
+   function is a reasonable threshold).
+3. Stub files for the extension modules (e.g. `torch._C`).
+4. Good type annotation test coverage, by using `check_untyped_defs=True` for
+   the test suite (or adding type annotations to tests).
```
Given the brevity, it may be beneficial to propose a target for the `[mypy]` section of `mypy.ini` to further clarify the vision.

ezyang commented:
This LGTM. @rgommers I'll let you make the small updates first? Another place these docs could live is in the wiki, where you don't have to go through landing cycle as is the case in the main repo.

rgommers commented:
> I'll let you make the small updates first?

yep

> Another place these docs could live is in the wiki, where you don't have to go through landing cycle as is the case in the main repo.

That makes sense to me. I only used this because `mypy-README` already existed. It looks a bit out of place at the root of the repo; happy to delete it completely in this PR and move the content to the wiki.

ezyang commented:
> That makes sense to me. I only used this because mypy-README already existed. It looks a bit out of place at the root of the repo; happy to delete it completely in this PR and move the content to the wiki.

SGTM!

rgommers committed:
Update mypy-README

rgommers committed:
Update mypy.ini and mypy test to add --check-untyped-defs

rgommers committed:
Finish task list in mypy-README.md

rgommers committed:
Revert change I didn't mean to commit (build workaround)

rgommers committed:
Merge branch 'master' into mypy-test

rgommers committed:
Minor fixes

rgommers committed:
Add sparse_coo to `torch/__init__.pyi`, and fix one mistake

rgommers commented:
While finishing the "task list" for completing the type annotations I found it was better to add `--check-untyped-defs` to `mypy.ini` and the test suite now, so I changed the scope of this PR. There were another ~500 errors that that uncovered, so I added a bunch more `ignore_errors = True`.

The guide of how to add type annotations and ~180 tasks for it are now in https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch. Please have a look at that as well if you see any fundamental problem.

I think when this PR is merged we're good to start tackling those tasks.
